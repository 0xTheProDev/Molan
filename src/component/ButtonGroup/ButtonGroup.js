// @flow
import React, { PureComponent } from 'react';
import { Button, Icon } from 'semantic-ui-react';
import saveAs from 'save-as';

import template from 'util/template';

import Styles from './ButtonGroup.module.css';

type ButtonGroupProps = {
  lang: string,
  code: string,
  onChange: (text: string) => void,
  onReload: () => void,
};

export class ButtonGroup extends PureComponent {
  fileBtn: ?React.Node;

  constructor(props: ButtonGroupProps) {
    super(props);
    this.fileBtn = null;
  }

  onFileBtnClick = () => {
    this.fileBtn.click();
  };

  onFileBtnChange = (e) => {
    const { onChange } = this.props;
    const fileReader = new FileReader();
    fileReader.onload = function (f) {
      onChange(f.target.result);
    }
    fileReader.readAsText(e.target.files[0], 'utf-8');
  };

  onSaveBtnClick = () => {
    const { lang, code } = this.props;
    const blob = new Blob([code], { type: 'text/plain;charset=ascii' });
    const item = template.find(e => e.lang === lang);
    saveAs(blob, `molan${item ? item.ext : '.txt'}`);
  };

  render() {
    const { onReload } = this.props;

    return (
      <React.Fragment>
        <Button animated="fade" floated="right" onClick={this.onSaveBtnClick}>
          <Button.Content hidden>Save</Button.Content>
          <Button.Content visible>
            <Icon name="save" />
          </Button.Content>
        </Button>
        <Button animated floated="right" onClick={onReload}>
          <Button.Content hidden>Reload</Button.Content>
          <Button.Content visible>
            <Icon name="redo" />
          </Button.Content>
        </Button>
        <Button animated="vertical" floated="right" onClick={this.onFileBtnClick}>
          <Button.Content hidden>Upload</Button.Content>
          <Button.Content visible>
            <Icon name="upload" />
          </Button.Content>
          <input
            type="file"
            accepts=".c,.cpp,.java,.js,.py"
            ref={input => this.fileBtn = input}
            onChange={this.onFileBtnChange}
            className={Styles['file-input']}
          />
        </Button>
      </React.Fragment>
    );
  }
}
