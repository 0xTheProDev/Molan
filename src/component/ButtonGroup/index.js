import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, Icon } from 'semantic-ui-react';
import saveAs from 'save-as';
import template from 'util/template';

export default class ButtonGroup extends Component {
    static propTypes = {
        lang:     PropTypes.string.isRequired,
        code:     PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired,
        onReload: PropTypes.func.isRequired
    };

    constructor(props) {
        super(props);
        this.fileBtn = null;
    }

    onFileBtnClick = () => {
        this.fileBtn.click();
    };

    onFileBtnChange = (e) => {
        const { onChange } = this.props;
        const fileReader = new FileReader();
        fileReader.onload = function(f) {
            onChange(f.target.result);
        }
        fileReader.readAsText(e.target.files[0], 'utf-8');
    };

    onSaveBtnClick = () => {
        const { lang, code } = this.props;
        const blob = new Blob([ code ], { type: 'text/plain;charset=ascii' });
        const item = template.find(e => e.lang === lang);
        saveAs(blob, `molan${item ? item.ext : '.txt'}`);
    };

    render() {
        const { onReload } = this.props;

        return (
            <div>
                <Button animated='fade' floated='right' onClick={this.onSaveBtnClick}>
                    <Button.Content hidden>Save</Button.Content>
                    <Button.Content visible>
                        <Icon name='save'/>
                    </Button.Content>
                </Button>
                <Button animated floated='right' onClick={() => onReload()}>
                    <Button.Content hidden>Reload</Button.Content>
                    <Button.Content visible>
                        <Icon name='redo'/>
                    </Button.Content>
                </Button>
                <Button animated='vertical' floated='right' onClick={this.onFileBtnClick}>
                    <Button.Content hidden>Upload</Button.Content>
                    <Button.Content visible>
                        <Icon name='upload'/>
                    </Button.Content>
                    <input
                      type='file'
                      accepts=".c,.cpp,.java,.js,.py"
                      style={{ display: 'none' }}
                      ref={input => this.fileBtn = input}
                      onChange={this.onFileBtnChange}
                    />
                </Button>
            </div>
        );
    }
}