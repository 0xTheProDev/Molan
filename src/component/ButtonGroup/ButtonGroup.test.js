import React from 'react';

import { ButtonGroup } from './ButtonGroup';

describe('ButtonGroup', () => {
  let
    onChangeStub,
    onResetStub,
    minProps;

  beforeAll(() => {
    onChangeStub = jest.fn();
    onResetStub = jest.fn();

    minProps = {
      lang: 'c',
      code: 'int main(void)',
      onChange: onChangeStub,
      onReset: onResetStub,
    };
  });

  it('should render', () => {
    const wrapper = shallow(<ButtonGroup {...minProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
