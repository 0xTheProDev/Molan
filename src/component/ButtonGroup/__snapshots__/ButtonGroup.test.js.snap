// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render 1`] = `
<Fragment>
  <Button
    animated="fade"
    as="button"
    floated="right"
    onClick={[Function]}
  >
    <ButtonContent
      hidden={true}
    >
      Save
    </ButtonContent>
    <ButtonContent
      visible={true}
    >
      <Icon
        as="i"
        name="save"
      />
    </ButtonContent>
  </Button>
  <Button
    animated={true}
    as="button"
    floated="right"
  >
    <ButtonContent
      hidden={true}
    >
      Reload
    </ButtonContent>
    <ButtonContent
      visible={true}
    >
      <Icon
        as="i"
        name="redo"
      />
    </ButtonContent>
  </Button>
  <Button
    animated="vertical"
    as="button"
    floated="right"
    onClick={[Function]}
  >
    <ButtonContent
      hidden={true}
    >
      Upload
    </ButtonContent>
    <ButtonContent
      visible={true}
    >
      <Icon
        as="i"
        name="upload"
      />
    </ButtonContent>
    <input
      accepts=".c,.cpp,.java,.js,.py"
      className="file-input"
      onChange={[Function]}
      type="file"
    />
  </Button>
</Fragment>
`;
